src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Emitter.cs

```diff
@@ -2,1289 +2,441 @@
-                EmitGetCoreMethod();
-                EmitGetValueCoreMethod();
-                EmitBindCoreMethods();
-                EmitInitializeMethods();
-                EmitHelperMethods();
+        public Emitter(SourceProductionContext context, SourceGenSpec sourceGenSpec)
+        {
+            _context = context;
+            SourceGenSpec = sourceGenSpec;
+        }
 
-                _writer.WriteBlockEnd(); // End helper class.
-                _writer.WriteBlockEnd(); // End containing namespace.
-                #endregion
+        public enum InitializationKind
+        {
+            None = 0,
+            SimpleAssignment = 1,
+            AssignmentWithNullCheck = 2,
+            Declaration = 3,
+        }
 
-                _context.AddSource($"{Identifier.GeneratedConfigurationBinder}.g.cs", _writer.ToSourceText());
-            }
+        public SourceGenSpec SourceGenSpec { get; }
 
-            #region Generated binder for user consumption.
-            private void EmitConfigureMethod()
-            {
-                if (!_generationSpec.ShouldEmitMethods(BinderMethodSpecifier.Configure))
-                {
-                    return;
-                }
+        public SourceWriter Writer { get; } = new();
 
-                _writer.WriteBlockStart($"public static {FullyQualifiedDisplayName.IServiceCollection} {Identifier.Configure}<T>(this {FullyQualifiedDisplayName.IServiceCollection} {Identifier.services}, {FullyQualifiedDisplayName.IConfiguration} {Identifier.configuration})");
+        public bool PrecedingBlockExists { get; set; }
 
-                EmitCheckForNullArgument_WithBlankLine(Identifier.configuration, useFullyQualifiedNames: true);
+        public bool UseFullyQualifiedNames { get; private set; }
 
-                foreach (TypeSpec type in _generationSpec.ConfigTypes[BinderMethodSpecifier.Configure])
-                {
-                    string typeDisplayString = type.FullyQualifiedDisplayString;
+        public void Emit()
+        {
+            Debug.Assert(SourceGenSpec.HasRootMethods());
 
-                    _writer.WriteBlockStart($"if (typeof(T) == typeof({typeDisplayString}))");
+            Writer.WriteLine(@"// <auto-generated/>
+#nullable enable
+");
+            UseFullyQualifiedNames = true;
+            SourceGenSpec.ConfigBinderSpec.Emit(this);
+            SourceGenSpec.OptionsBuilderSpec.Emit(this);
+            SourceGenSpec.ServiceCollectionSpec.Emit(this);
 
-                    _writer.WriteBlockStart($@"return {Identifier.services}.{Identifier.Configure}<{typeDisplayString}>({Identifier.obj} =>");
-                    EmitIConfigurationHasValueOrChildrenCheck(voidReturn: true);
-                    EmitBindLogicFromRootMethod(type, Identifier.obj, InitializationKind.None);
-                    _writer.WriteBlockEnd(");");
+            UseFullyQualifiedNames = false;
+            SourceGenSpec.CoreBindingHelperSpec.Emit(this);
```

Comment:
This is the major diff in this file
- generation for `ConfigurationBinder` moved to its own file
- each extension class has it's own emitter file
- core binding logic shared by each extension class also moved to a dedicated file.